### PR TITLE
Add support for cubemap arrays in the backends

### PR DESCRIPTION
**Pull request by pixelflinger**
This is not intended to be used yet because we're not currently
checking that cubemap array are actually supported, however, if they
are, they should work.

### CODE REVIEW

File: filament/backend/src/metal/MetalHandles.mm [hunk 1]
```diff
@@ -433,46 +433,38 @@ BufferDescriptor b(nullptr, 0u, [](void* buffer, size_t size, void* user) {
...
-                return MTLTextureType2D;
-            case 0b01:
-                return MTLTextureType2DArray;
-            case 0b10:
-                return MTLTextureType2DMultisample;
-            case 0b11:
+    const auto get2DTextureType = [](SamplerType target, bool isMultisampled) {
+        if (!isMultisampled) {
+            return getMetalType(target);
+        } else {
```
Comment (bejado) [inline filament/backend/src/metal/MetalHandles.mm:445]: You can remove the `else` statement here.

File: filament/backend/src/metal/MetalHandles.mm [hunk 2]
```diff
@@ -433,46 +433,38 @@ BufferDescriptor b(nullptr, 0u, [](void* buffer, size_t size, void* user) {
...
+                    return getMetalType(target);
+            }
         }
     };
 
     MTLTextureDescriptor* descriptor;
     switch (target) {
         case SamplerType::SAMPLER_2D:
         case SamplerType::SAMPLER_2D_ARRAY:
+        case SamplerType::SAMPLER_CUBEMAP_ARRAY:
```
Comment (bejado) [inline filament/backend/src/metal/MetalHandles.mm:464]: I think we can put this `case` statement next to the existing `case SamplerType::SAMPLER_CUBEMAP` on line 478 by copying down the `descriptor.arrayLength = depth` line.

File: filament/backend/src/metal/MetalHandles.mm (hunk 1 above)
Comment (pixelflinger) [reply filament/backend/src/metal/MetalHandles.mm]: done

File: filament/backend/src/metal/MetalHandles.mm (hunk 2 above)
Comment (pixelflinger) [reply filament/backend/src/metal/MetalHandles.mm]: done -- and also I think my version didn't work because the texture type was wrong.

**Comment by bejado** [comment]
LGTM, but iOS is failing because it doesn't have 3.1 headers:

```
/Users/runner/work/filament/filament/filament/backend/src/opengl/OpenGLContext.h:408:14: error: use of undeclared identifier 'GL_TEXTURE_CUBE_MAP_ARRAY'
        case GL_TEXTURE_CUBE_MAP_ARRAY:         return 6;
```

I can help fix.

**Comment by pixelflinger** [comment]
oh... I lost your fix. damn.

**Comment by bejado** [comment]
> oh... I lost your fix. damn.

Here's the commit, you can cherry-pick it to your branch: https://github.com/google/filament/commit/66ab68aeb74dd15229feb43c4a269c06f341515e

**Comment by bejado** [comment]
Took another look at the Metal changes, LGTM.

**Review by bejado — APPROVED** [review]

**Pull request merged**